File: src/HolzShots/Input/Actions/WindowCommand.vb
Imports HolzShots.Composition.Command
Imports HolzShots.Threading
Imports HolzShots.Drawing
Imports System.Drawing.Drawing2D
Imports System.Drawing.Imaging
Imports HolzShots.Windows.Forms

Namespace Input.Actions

    <Command("captureWindow")>
    Public Class WindowCommand
        Inherits CapturingCommand

        Public Overrides Async Function Invoke(parameters As IReadOnlyDictionary(Of String, String), settingsContext As HSSettings) As Task
            ' TODO: Add proper assertion
            ' Debug.Assert(ManagedSettings.EnableWindowScreenshot)

            ' TODO: Re-add proper if condition
            ' If ManagedSettings.EnableWindowScreenshot Then
            Dim h As IntPtr = Native.User32.GetForegroundWindow()

            Dim info As Native.User32.WindowPlacement
            Native.User32.GetWindowPlacement(h, info)

            Dim shot = CaptureWindow(h)
            Await ProcessCapturing(shot, settingsContext).ConfigureAwait(True)
            ' End If
        End Function

        Private Shared Function CaptureWindow(windowHandle As IntPtr, Optional includeMargin As Boolean = True) As Screenshot
            If Native.User32.IsIconic(windowHandle) Then Return Nothing

            Using prio As New ProcessPriorityRequest()
                Using shotSet = GetShotSet(windowHandle, includeMargin)
                    Return Screenshot.FromWindow(shotSet)
                End Using
            End Using
        End Function

        Private Shared Function GetShotSet(windowHandle As IntPtr, includeMargin As Boolean) As WindowScreenshotSet
            ' TODO: Refactor methods to WindowScreenshotSet?
            If HolzShots.Windows.Forms.EnvironmentEx.IsAeroEnabled() Then
                Return DoAeroOn(windowHandle, includeMargin, False)
            ElseIf HolzShots.Windows.Forms.EnvironmentEx.IsVistaOrHigher Then
                Return DoAeroOff(windowHandle)
            Else
                Debugger.Break() ' wait, you prick!
                Throw New InvalidOperationException("Unsupported operating system.")
            End If
        End Function

        ' TODO: Rewrite this whole mess

        Private Shared Function DoAeroOn(wndHandle As IntPtr, includeMargin As Boolean, smallMargin As Boolean) As WindowScreenshotSet

            Dim nativeRectangle As Native.Rect
            Native.User32.GetWindowRect(wndHandle, nativeRectangle)

            Dim placement As Native.User32.WindowPlacement
            Native.User32.GetWindowPlacement(wndHandle, placement)

            If includeMargin Then
                If placement.showCmd <> 3 Then
                    Dim left = nativeRectangle.Left - If(smallMargin, 4, 17)
                    Dim top = nativeRectangle.Top - If(smallMargin, 4, 17)
                    Dim right = nativeRectangle.Right + If(smallMargin, 4, 21)
                    Dim bottom = nativeRectangle.Bottom + If(smallMargin, 4, 21)

                    nativeRectangle = New Native.Rect(
                        Math.Max(left, SystemInformation.VirtualScreen.Left),
                        Math.Max(top, SystemInformation.VirtualScreen.Top),
                        Math.Min(right, SystemInformation.VirtualScreen.Right),
                        Math.Min(bottom, SystemInformation.VirtualScreen.Bottom)
                    )
                Else
                    Dim tmprect As Rectangle = nativeRectangle
                    Dim center As New Point(tmprect.X + CInt(tmprect.Width / 2), tmprect.Y + CInt(tmprect.Height / 2))
                    nativeRectangle = Screen.GetWorkingArea(center) 'NativeTypes.Rect.FromRectangle(Screen.GetWorkingArea(center))
                End If
            End If

            Dim drawingRectangle As Rectangle = nativeRectangle

            If drawingRectangle.Size.Height < 0 OrElse drawingRectangle.Size.Width < 0 Then Return Nothing

            Dim cursorPositonOnScreenshot As New Point(Cursor.Position.X - drawingRectangle.Location.X, Cursor.Position.Y - drawingRectangle.Location.Y)


            Using bg As New BackgroundForm(drawingRectangle.Location, drawingRectangle.Size) _
                'New Point(rct.Left, rct.Top), New Size(rct.Right - rct.Left, rct.Bottom - rct.Top))
                'Using bg As New FloatingWindow(nrct.X, nrct.Y, nrct.Width, nrct.Height)
                Using bmpBlack As New Bitmap(drawingRectangle.Width, drawingRectangle.Height, PixelFormat.Format32bppPArgb)
                    Using bmpWhite As New Bitmap(drawingRectangle.Width, drawingRectangle.Height, PixelFormat.Format32bppPArgb)

                        bg.Visible = True

                        WindowRedraw.StopRedraw(wndHandle)

                        Native.User32.SetForegroundWindowEx(bg.Handle)
                        Native.User32.SetForegroundWindowEx(wndHandle)

                        Using ga As Graphics = Graphics.FromImage(bmpBlack)
                            ga.CompositingQuality = CompositingQuality.HighQuality
                            ga.CopyFromScreen(drawingRectangle.X, drawingRectangle.Y, 0, 0, bg.Size)
                        End Using

                        bg.BackColor = Color.White
                        bg.Refresh()

                        Using ga As Graphics = Graphics.FromImage(bmpWhite)
                            ga.CopyFromScreen(drawingRectangle.X, drawingRectangle.Y, 0, 0, bg.Size)
                        End Using

                        WindowRedraw.StartRedraw(wndHandle)

                        bg.Visible = False

                        Dim result As New Bitmap(bmpWhite.Width, bmpWhite.Height)

                        Computation.ComputeAlphaChannel(bmpWhite, bmpBlack, result)

                        ' Old method:
                        ' ScreenshotMethodsHelper.ComputeAlphaChannel(bmpWhite, bmpBlack)

                        Dim windowTitle = WindowInformation.GetWindowTitle(wndHandle)
                        Dim processName = WindowInformation.GetProcessNameOfWindow(wndHandle)

                        Return New WindowScreenshotSet(result, cursorPositonOnScreenshot, windowTitle, processName)
                    End Using
                End Using
            End Using
        End Function

        Private Shared Function DoAeroOff(wndHandle As IntPtr) As WindowScreenshotSet
            Dim nativeRectangle As Native.Rect
            Native.User32.GetWindowRect(wndHandle, nativeRectangle)
            Dim drawingRectangle As Rectangle = nativeRectangle

            Dim cursorPositonOnScreenshot As New Point(Cursor.Position.X - drawingRectangle.Location.X, Cursor.Position.Y - drawingRectangle.Location.Y)

            Dim bmp = ScreenshotCreator.CaptureScreenshot(drawingRectangle)

            Dim windowTitle = WindowInformation.GetWindowTitle(wndHandle)
            Dim processName = WindowInformation.GetProcessNameOfWindow(wndHandle)

            Return New WindowScreenshotSet(bmp, cursorPositonOnScreenshot, windowTitle, processName)
        End Function

    End Class
End Namespace

File: src/HolzShots/Input/Action/WindowCommand.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.Threading.Tasks;
using System.Windows.Forms;
using HolzShots.Composition.Command;
using HolzShots.Drawing;
using HolzShots.Threading;
using HolzShots.Windows.Forms;

namespace HolzShots.Input.Actions
{
    [Command("captureWindow")]
    public class WindowCommand : CapturingCommand
    {
        protected override async Task InvokeInternal(IReadOnlyDictionary<string, string> parameters, HSSettings settingsContext)
        {
            // TODO: Add proper assertion
            // Debug.Assert(ManagedSettings.EnableWindowScreenshot)

            // TODO: Re-add proper if condition
            // If ManagedSettings.EnableWindowScreenshot Then
            var h = Native.User32.GetForegroundWindow();

            Native.User32.GetWindowPlacement(h, out Native.User32.WindowPlacement info);

            var shot = CaptureWindow(h);
            await ProcessCapturing(shot, settingsContext).ConfigureAwait(true);
        }

        private static Screenshot CaptureWindow(IntPtr windowHandle, bool includeMargin = true)
        {
            if (Native.User32.IsIconic(windowHandle))
                return default;

            using var prio = new ProcessPriorityRequest();
            using var shotSet = GetShotSet(windowHandle, includeMargin);
            return Screenshot.FromWindow(shotSet);
        }

        private static WindowScreenshotSet GetShotSet(IntPtr windowHandle, bool includeMargin)
        {
            // TODO: Refactor methods to WindowScreenshotSet?
            if (HolzShots.Windows.Forms.EnvironmentEx.IsAeroEnabled())
                return DoAeroOn(windowHandle, includeMargin, false);
            else if (HolzShots.Windows.Forms.EnvironmentEx.IsVistaOrHigher)
                return DoAeroOff(windowHandle);
            else
            {
                Debugger.Break(); // wait, you prick!
                throw new InvalidOperationException("Unsupported operating system.");
            }
        }

        // TODO: Rewrite this whole mess

        private static WindowScreenshotSet DoAeroOn(IntPtr wndHandle, bool includeMargin, bool smallMargin)
        {
            Native.User32.GetWindowRect(wndHandle, out Native.Rect nativeRectangle);

            Native.User32.GetWindowPlacement(wndHandle, out Native.User32.WindowPlacement placement);

            if (includeMargin)
            {
                if (placement.showCmd != 3)
                {
                    var left = nativeRectangle.Left - (smallMargin ? 4 : 17);
                    var top = nativeRectangle.Top - (smallMargin ? 4 : 17);
                    var right = nativeRectangle.Right + (smallMargin ? 4 : 21);
                    var bottom = nativeRectangle.Bottom + (smallMargin ? 4 : 21);

                    nativeRectangle = new Native.Rect(Math.Max(left, SystemInformation.VirtualScreen.Left), Math.Max(top, SystemInformation.VirtualScreen.Top), Math.Min(right, SystemInformation.VirtualScreen.Right), Math.Min(bottom, SystemInformation.VirtualScreen.Bottom)
);
                }
                else
                {
                    Rectangle tmprect = nativeRectangle;
                    Point center = new Point(tmprect.X + System.Convert.ToInt32(tmprect.Width / (double)2), tmprect.Y + System.Convert.ToInt32(tmprect.Height / (double)2));
                    nativeRectangle = Screen.GetWorkingArea(center); // NativeTypes.Rect.FromRectangle(Screen.GetWorkingArea(center))
                }
            }

            Rectangle drawingRectangle = nativeRectangle;

            if (drawingRectangle.Size.Height < 0 || drawingRectangle.Size.Width < 0)
                return default;

            var cursorPositonOnScreenshot = new Point(Cursor.Position.X - drawingRectangle.Location.X, Cursor.Position.Y - drawingRectangle.Location.Y);


            using (BackgroundForm bg = new BackgroundForm(drawingRectangle.Location, drawingRectangle.Size))
            {
                // New Point(rct.Left, rct.Top), New Size(rct.Right - rct.Left, rct.Bottom - rct.Top))
                // Using bg As New FloatingWindow(nrct.X, nrct.Y, nrct.Width, nrct.Height)
                using (Bitmap bmpBlack = new Bitmap(drawingRectangle.Width, drawingRectangle.Height, PixelFormat.Format32bppPArgb))
                {
                    using (Bitmap bmpWhite = new Bitmap(drawingRectangle.Width, drawingRectangle.Height, PixelFormat.Format32bppPArgb))
                    {
                        bg.Visible = true;

                        WindowRedraw.StopRedraw(wndHandle);

                        Native.User32.SetForegroundWindowEx(bg.Handle);
                        Native.User32.SetForegroundWindowEx(wndHandle);

                        using (var ga = Graphics.FromImage(bmpBlack))
                        {
                            ga.CompositingQuality = CompositingQuality.HighQuality;
                            ga.CopyFromScreen(drawingRectangle.X, drawingRectangle.Y, 0, 0, bg.Size);
                        }

                        bg.BackColor = Color.White;
                        bg.Refresh();

                        using (var ga = Graphics.FromImage(bmpWhite))
                        {
                            ga.CopyFromScreen(drawingRectangle.X, drawingRectangle.Y, 0, 0, bg.Size);
                        }

                        WindowRedraw.StartRedraw(wndHandle);

                        bg.Visible = false;

                        var result = new Bitmap(bmpWhite.Width, bmpWhite.Height);
                        Computation.ComputeAlphaChannel(bmpWhite, bmpBlack, ref result);

                        // Old method:
                        // ScreenshotMethodsHelper.ComputeAlphaChannel(bmpWhite, bmpBlack)

                        var windowTitle = WindowInformation.GetWindowTitle(wndHandle);
                        var processName = WindowInformation.GetProcessNameOfWindow(wndHandle);

                        return new WindowScreenshotSet(result, cursorPositonOnScreenshot, windowTitle, processName);
                    }
                }
            }
        }

        private static WindowScreenshotSet DoAeroOff(IntPtr wndHandle)
        {
            Native.User32.GetWindowRect(wndHandle, out Native.Rect nativeRectangle);
            Rectangle drawingRectangle = nativeRectangle;

            var cursorPositonOnScreenshot = new Point(Cursor.Position.X - drawingRectangle.Location.X, Cursor.Position.Y - drawingRectangle.Location.Y);

            var bmp = ScreenshotCreator.CaptureScreenshot(drawingRectangle);

            var windowTitle = WindowInformation.GetWindowTitle(wndHandle);
            var processName = WindowInformation.GetProcessNameOfWindow(wndHandle);

            return new WindowScreenshotSet(bmp, cursorPositonOnScreenshot, windowTitle, processName);
        }
    }
}
